### PR TITLE
Add ATF/FIP and OP-TEE into imx6ul image

### PR DESCRIPTION
**Pull request by bryanodonoghue**
Adds.
1. ATF
Because we need to include code from u-boot SPL to initialize DDR the final ATF patch has been split out and added into the meta-mbl repo for application during build.

https://git.linaro.org/landing-teams/working/mbl/arm-trusted-firmware.git/commit/?h=linaro-imx6-imx7-imx8-noddr&id=e39ac9f802da8649ed1b163c95273b930508bb90

Here is how the tree should look with the DDR init included
https://git.linaro.org/landing-teams/working/mbl/arm-trusted-firmware.git/commit/?h=linaro-imx6-imx7-imx8&id=d3f24d80721e92b98ddf5b3dcb85fdef169ed54d

2. FIP
Normal FIP as we have on other platforms.

3. u-boot in BL33 mode
A normal implementation of BL33 u-boot
https://git.linaro.org/landing-teams/working/mbl/u-boot.git/commit/?h=linaro-imx&id=2ccdb910c023a691ac00ed40b3f3aacf45c82cde

4. OP-TEE
A straight-forward OP-TEE port, just a few configuration parameters and aprox < ~ 10 LOC
https://git.linaro.org/landing-teams/working/mbl/optee_os.git/commit/?h=linaro-main&id=476fc86142f058c7190733243f48c5c0115157e9

I still have to add OPTEE client to the image but, this is enough to get the ball rolling on integration, test and review.

### CODE REVIEW

File: meta-freescale-3rdparty-mbl/conf/machine/imx6ul-pico-mbl.conf [hunk 1]
```diff
@@ -122,13 +121,12 @@ MACHINE_EXTRA_RRECOMMENDS_append = " \
...
+MBL_WKS_BOOTLOADER1_FILENAME = "bl2.bin.imx"
 MBL_WKS_BOOTLOADER1_OFFSET_BANK1_KiB = "1"
-MBL_WKS_BOOTLOADER1_SIZE_KiB = "68"
+MBL_WKS_BOOTLOADER1_SIZE_KiB = "1023"
 
-MBL_WKS_BOOTLOADER2_FILENAME = "u-boot.img"
-MBL_WKS_BOOTLOADER2_ALIGN_KiB = "1"
-MBL_WKS_BOOTLOADER2_OFFSET_BANK1_KiB = "69"
+MBL_WKS_BOOTLOADER2_OFFSET_BANK1_KiB = "1024"
+MBL_WKS_BOOTLOADER2_ALIGN_KiB = "1024"
```
Comment (diego-sueiro) [inline meta-freescale-3rdparty-mbl/conf/machine/imx6ul-pico-mbl.conf]: I suppose we need a similar change as in the commit f749a2264aedcbaf1dabe39679184f489dd49ec1.
@jonathanhaigh-arm, can you please confirm?

File: meta-freescale-mbl/recipes-bsp/u-boot/u-boot-fslc_%.bbappend [hunk 2]
```diff
@@ -32,9 +29,9 @@ DCD_FILE_PATH_imx7d-pico-mbl = "${B}"
 DCD_FILE_PATH_imx6ul-pico-mbl = "${B}"
 DCD_FILE_PATH_imx6ul-des0258-mbl = "${B}"
 
-do_deploy_prepend_imx6ul-pico-mbl() {
-	cp ${B}/spl/u-boot-spl.cfgout ${B}/u-boot-dtb.cfgout
-}
+#do_deploy_prepend_imx6ul-pico-mbl() {
```
Comment (diego-sueiro) [inline meta-freescale-mbl/recipes-bsp/u-boot/u-boot-fslc_%.bbappend]: Can you please remove the commented lines?

File: meta-freescale-3rdparty-mbl/recipes-bsp/atf/atf-imx6ul-pico-mbl_git.bb [hunk 3]
```diff
@@ -2,15 +2,8 @@
 #
 # SPDX-License-Identifier: MIT
 
-inherit mbl-artifact-names
-require recipes-bsp/atf/atf.inc
+require atf-imx7-mbl_git.inc
```
Comment (diego-sueiro) [inline meta-freescale-3rdparty-mbl/recipes-bsp/atf/atf-imx6ul-pico-mbl_git.bb]: Should we rename the `atf-imx7-mbl_git.inc` file to `atf-imx-mbl_git.inc` as soon as it is now being used by imx7 and imx6 platforms?

File: meta-mbl-bsp-common/recipes-bsp/atf/atf.inc [hunk 4]
```diff
@@ -9,14 +9,17 @@ DEPENDS += " openssl-native coreutils-native optee-os u-boot virtual/kernel"
 
 PROVIDES += "virtual/atf"
 
+FILESEXTRAPATHS_prepend:="${THISDIR}/files:"
+
 SRC_URI = " \
-    git://git.linaro.org/landing-teams/working/mbl/arm-trusted-firmware.git;protocol=https;nobranch=1;name=atf \
-    git://github.com/ARMmbed/mbedtls.git;protocol=https;branch=development;name=mbedtls;destsuffix=mbedtls \
+    git://git.linaro.org/landing-teams/working/mbl/arm-trusted-firmware.git;protocol=https;nobranch=1;name=atf	\
+    file://0001-imx-imx6-uboot_ddr_init-Import-u-boot-DDR-init-code.patch					\
```
Comment (diego-sueiro) [inline meta-mbl-bsp-common/recipes-bsp/atf/atf.inc]: This patch should be added in `meta-freescale-3rdparty-mbl/recipes-bsp/atf/atf-imx6ul-pico-mbl_git.bb` instead. Both the `SRC_URI` and the patch file itself (in `meta-freescale-3rdparty-mbl/recipes-bsp/atf/atf-imx6ul-pico-mbl` directory)

File: meta-freescale-3rdparty-mbl/recipes-bsp/atf/atf-imx6ul-pico-mbl_git.bb [hunk 5]
```diff
@@ -2,15 +2,8 @@
...
-# to here: with atf-imx7-mbl_git.inc if/when compiling ATF for this target
+PLATFORM = "imx6ul_picopi"
 
-PLATFORM = "picopi"
-
-ATF_COMPILE_FLAGS = " "
-
-do_compile[noexec] = "1"
-do_deploy[noexec] = "1"
+ATF_COMPILE_FLAGS += "BUILD_UBOOT_DRAM_INIT=1"
```
Comment (diego-sueiro) [inline meta-freescale-3rdparty-mbl/recipes-bsp/atf/atf-imx6ul-pico-mbl_git.bb]: Can you please add a comment about this addition?

File: meta-mbl-bsp-common/recipes-bsp/atf/files/0001-imx-imx6-uboot_ddr_init-Import-u-boot-DDR-init-code.patch [hunk 6]
```diff
@@ -0,0 +1,1641 @@
...
+Subject: [PATCH] imx: imx6: uboot_ddr_init: Import u-boot DDR init code
+
+Import u-boot DDR init code for imx6ul. Since the u-boot stuff is GPL-2.0+
+we can't send it to the ATF repo. Therefore this patch must not be included
+in any PR to ATF.
+
+It should probably be a standalone patch in the meta-mbl layer and applied
+during building of the i.MX6 ATF binary.
+
+For now just stack it as a patch for the Linaro landing area to host.
```
Comment (diego-sueiro) [inline meta-mbl-bsp-common/recipes-bsp/atf/files/0001-imx-imx6-uboot_ddr_init-Import-u-boot-DDR-init-code.patch]: Can you please specify from where this code comes from (repo URL and sha)?

File: meta-freescale-3rdparty-mbl/recipes-bsp/atf/atf-imx6ul-pico-mbl_git.bb (hunk 3 above)
Comment (bryanodonoghue) [reply meta-freescale-3rdparty-mbl/recipes-bsp/atf/atf-imx6ul-pico-mbl_git.bb]: atf-imx6-imx7-mbl_git.inc since it is not used on imx8

**Comment by bryanodonoghue** [comment]
Right now xtest runs but doesn't pass.

I think it is worth trying to merge as is and dealing with xtest in isolation.

root@mbed-linux-os-5665:~# xtest
Run test suite with level=0[   93.982572] Bad mode in prefetch abort handler detected
[   93.988549] Internal error: Oops - bad mode: 0 [#1] SMP ARM
[   93.994144] Modules linked in: usb_f_ecm g_ether usb_f_rndis u_ether libcomposite configfs brcmfmac brcmutil evbug
[   94.004606] CPU: 0 PID: 290 Comm: xtest Not tainted 4.14.112-fslc+g45bad79 #1
[   94.011759] Hardware name: Freescale i.MX6 Ultralite (Device Tree)
[   94.017960] task: d70a0000 task.stack: d6f8a000
[   94.022513] PC is at 0x9e800728
[   94.025685] LR is at optee_smccc_smc+0x40/0x48
[   94.030150] pc : [<9e800728>]    lr : [<c079edf4>]    psr: 600701d6
[   94.036435] sp : d6f8bce8  ip : d6f8bd40  fp : d6f8bd6c
[   94.041681] r10: d708e680  r9 : d708e840  r8 : d6a5da0c
[   94.046925] r7 : 00000000  r6 : 00000000  r5 : 00000000  r4 : 00000000
[   94.053473] r3 : 00000000  r2 : 9fe00000  r1 : 00000000  r0 : 32000004
[   94.060023] Flags: nZCv  IRQs off  FIQs off  Mode MON_32  ISA ARM  Segment none
[   94.067353] Control: 10c5387d  Table: 970ac06a  DAC: 00000000
[   94.073119] Process xtest (pid: 290, stack limit = 0xd6f8a210)
[   94.078975] Stack: (0xd6f8bce8 to 0xd6f8c000)
[   94.083362] bce0:                   32000004 00000000 9fe00000 00000000 00000000 00000000
[   94.091567] bd00: 00000000 00000000 d6a5da0c d708e840 d708e680 d6f8bd6c d6f8bd40 d6f8bce8
[   94.099772] bd20: c079edf4 9e800728 600701d6 ffffffff 00000000 d6a5da00 d6f8bd8c ffff0000
[   94.107976] bd40: 00000000 00000000 00000000 00000000 d6f8bd8c 00000000 9fe00000 c079edb4
[   94.116181] bd60: d6f8be14 d6f8bd70 c079f44c c079edc0 00000000 00000000 00000000 00000000
[   94.124387] bd80: d6f8bd8c 00000000 00000000 00000000 014080c0 d6001e00 014080c0 32000004
[   94.132591] bda0: 00000000 9fe00000 00000000 00000000 00000000 00000000 00000000 d6a5da50
[   94.140795] bdc0: d6a5da50 00000000 00000000 dead4ead ffffffff ffffffff c1aed364 00000000
[   94.149000] bde0: 00000000 c0d014ac d6f8bde8 d6f8bde8 d6f8bea8 d708e4c0 00000000 d708e840
[   94.157204] be00: d7028580 d7028280 d6f8be4c d6f8be18 c079f6ac c079f3bc dea80000 9fe00000
[   94.165409] be20: d6f8be6c d708e840 00000000 be8c1a80 d7028580 d6f8bea8 00000000 00000000
[   94.173614] be40: d6f8bf0c d6f8be50 c079d410 c079f5b4 00000022 00d1cc12 00000022 d7007140
[   94.181820] be60: d6f8be84 d6f8be70 c050cc40 c0179748 00000022 00000000 d6f8be94 d6f8be88
[   94.190026] be80: 00000080 00000000 be8c1ab8 d6f8be98 c0501ce4 c050a16c be8c1a80 00000000
[   94.198230] bea0: 000000b8 00000000 405b6ad9 e321e5c3 02109487 1bc6d5a5 00000000 00000000
[   94.206433] bec0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000004
[   94.214638] bee0: 000d23b8 be8c1bc8 d6ca9640 00000003 8010a402 c026b9d8 00000003 d6a7c830
[   94.222844] bf00: d6f8bf7c d6f8bf10 c026b000 c079ce2c d6ca9640 d6653000 00000020 00000000
[   94.231049] bf20: d7007140 d62417c0 00000002 00000022 d7007148 00000004 d6f8bf74 d6f8bf48
[   94.239253] bf40: c0256a5c c02a0d90 00000000 00000000 d7007140 d6ca9640 d6ca9640 00000003
[   94.247459] bf60: 8010a402 be8c1bc8 d6f8a000 00000036 d6f8bfa4 d6f8bf80 c026b9d8 c026af70
[   94.255665] bf80: 00000000 be8c1b30 00d16898 00000036 c0108044 d6f8a000 00000000 d6f8bfa8
[   94.263871] bfa0: c0107e60 c026b9a8 00000000 be8c1b30 00000003 8010a402 be8c1bc8 00000003
[   94.272076] bfc0: 00000000 be8c1b30 00d16898 00000036 be8c1e18 00000000 0014480c 000000bf
[   94.280280] bfe0: b6f9502c be8c1a74 b6f835e7 b6e982b8 60070030 00000003 00000000 00000000
[   94.288472] Backtrace: 
[   94.290959] [<c079edb4>] (optee_smccc_smc) from [<c079f44c>] (optee_do_call_with_arg+0x9c/0x110)
[   94.299765]  r4:c079edb4
[   94.302329] [<c079f3b0>] (optee_do_call_with_arg) from [<c079f6ac>] (optee_open_session+0x104/0x1e8)
[   94.311489]  r9:d7028280 r8:d7028580 r7:d708e840 r6:00000000 r5:d708e4c0 r4:d6f8bea8
[   94.319264] [<c079f5a8>] (optee_open_session) from [<c079d410>] (tee_ioctl+0x5f0/0x12d4)
[   94.327383]  r10:00000000 r9:00000000 r8:d6f8bea8 r7:d7028580 r6:be8c1a80 r5:00000000
[   94.335229]  r4:d708e840
[   94.337797] [<c079ce20>] (tee_ioctl) from [<c026b000>] (do_vfs_ioctl+0x9c/0xa38)
[   94.345222]  r10:d6a7c830 r9:00000003 r8:c026b9d8 r7:8010a402 r6:00000003 r5:d6ca9640
[   94.353069]  r4:be8c1bc8
[   94.355632] [<c026af64>] (do_vfs_ioctl) from [<c026b9d8>] (SyS_ioctl+0x3c/0x68)
[   94.362970]  r10:00000036 r9:d6f8a000 r8:be8c1bc8 r7:8010a402 r6:00000003 r5:d6ca9640
[   94.370817]  r4:d6ca9640
[   94.373385] [<c026b99c>] (SyS_ioctl) from [<c0107e60>] (ret_fast_syscall+0x0/0x28)
[   94.380982]  r9:d6f8a000 r8:c0108044 r7:00000036 r6:00d16898 r5:be8c1b30 r4:00000000
[   94.388752] Code: bad PC value
[   94.391836] ---[ end trace 15ac160c0d66c5e2 ]---


TEE test application started with device [(null)]
######################################################
#
# regression
#
######################################################
 
* regression_1001 Core self tests
Segmentation fault
root@mbed-linux-os-5665:~#

File: meta-mbl-bsp-common/recipes-bsp/atf/files/0001-imx-imx6-uboot_ddr_init-Import-u-boot-DDR-init-code.patch [hunk 7]
```diff
@@ -0,0 +1,1650 @@
+From 97faca4a323b99aa4d0c3c741ab0856f7a2e15a0 Mon Sep 17 00:00:00 2001
```
Comment (diego-sueiro) [inline meta-mbl-bsp-common/recipes-bsp/atf/files/0001-imx-imx6-uboot_ddr_init-Import-u-boot-DDR-init-code.patch]: This file should be moved to `meta-freescale-3rdparty-mbl/recipes-bsp/atf/files`

File: meta-mbl-bsp-common/recipes-bsp/atf/atf.inc [hunk 8]
```diff
@@ -9,14 +9,16 @@ DEPENDS += " openssl-native coreutils-native optee-os u-boot virtual/kernel"
 
 PROVIDES += "virtual/atf"
 
+FILESEXTRAPATHS_prepend:="${THISDIR}/files:"
```
Comment (diego-sueiro) [inline meta-mbl-bsp-common/recipes-bsp/atf/atf.inc]: By moving the patch to `meta-freescale-3rdparty` you have to move this line to `meta-freescale-3rdparty-mbl/recipes-bsp/atf/atf-imx6ul-pico-mbl_git.bb`.

File: meta-mbl-bsp-common/recipes-bsp/atf/atf.inc [hunk 9]
```diff
@@ -9,14 +9,16 @@ DEPENDS += " openssl-native coreutils-native optee-os u-boot virtual/kernel"
...
+
 SRC_URI = " \
-    git://git.linaro.org/landing-teams/working/mbl/arm-trusted-firmware.git;protocol=https;nobranch=1;name=atf \
-    git://github.com/ARMmbed/mbedtls.git;protocol=https;branch=development;name=mbedtls;destsuffix=mbedtls \
+    git://git.linaro.org/landing-teams/working/mbl/arm-trusted-firmware.git;protocol=https;nobranch=1;name=atf	\
+    git://github.com/ARMmbed/mbedtls.git;protocol=https;branch=development;name=mbedtls;destsuffix=mbedtls	\
```
Comment (diego-sueiro) [inline meta-mbl-bsp-common/recipes-bsp/atf/atf.inc]: Can you undo these space changes in the `SRC_URI`?

File: meta-freescale-3rdparty-mbl/recipes-bsp/atf/atf-imx6ul-pico-mbl_git.bb [hunk 10]
```diff
@@ -2,15 +2,15 @@
...
 
-ATF_COMPILE_FLAGS = " "
-
-do_compile[noexec] = "1"
-do_deploy[noexec] = "1"
+SRC_URI = " \
+    git://git.linaro.org/landing-teams/working/mbl/arm-trusted-firmware.git;protocol=https;nobranch=1;name=atf	\
+    file://0001-imx-imx6-uboot_ddr_init-Import-u-boot-DDR-init-code.patch					\
+    git://github.com/ARMmbed/mbedtls.git;protocol=https;branch=development;name=mbedtls;destsuffix=mbedtls	\
```
Comment (diego-sueiro) [inline meta-freescale-3rdparty-mbl/recipes-bsp/atf/atf-imx6ul-pico-mbl_git.bb]: `SRC_URI_append` didn't work?

**Comment by diego-sueiro** [comment]
> Right now xtest runs but doesn't pass.
> 
> I think it is worth trying to merge as is and dealing with xtest in isolation.

Ok. Can you please raise a JIRA ticket for this?

**Comment by bryanodonoghue** [comment]
> > Right now xtest runs but doesn't pass.
> > I think it is worth trying to merge as is and dealing with xtest in isolation.
> 
> Ok. Can you please raise a JIRA ticket for this?

https://jira.arm.com/browse/IOTMBL-2427

File: meta-freescale-3rdparty-mbl/recipes-bsp/atf/atf-imx6ul-pico-mbl_git.bb (hunk 10 above)
Comment (bryanodonoghue) [reply meta-freescale-3rdparty-mbl/recipes-bsp/atf/atf-imx6ul-pico-mbl_git.bb]: No unfortunately not.

**Comment by diego-sueiro** [comment]
> @bryanodonoghue,
> 
> Can you please add support for `FIP_ROM_OFFSET` for the pico6ul?
> References:
> [301a47b](https://github.com/ARMmbed/meta-mbl/commit/301a47b5c6cb2a5dafd4ff339a2a8e70e60ec9b2)
> https://git.linaro.org/landing-teams/working/mbl/arm-trusted-firmware.git/commit/?id=17e93a8fbf2e0925b23151fcb4a7cf1e2a1130f2

@bryanodonoghue, have you addressed this?

**Comment by bryanodonoghue** [comment]
> > @bryanodonoghue,
> > Can you please add support for `FIP_ROM_OFFSET` for the pico6ul?
> > References:
> > [301a47b](https://github.com/ARMmbed/meta-mbl/commit/301a47b5c6cb2a5dafd4ff339a2a8e70e60ec9b2)
> > https://git.linaro.org/landing-teams/working/mbl/arm-trusted-firmware.git/commit/?id=17e93a8fbf2e0925b23151fcb4a7cf1e2a1130f2
> 
> @bryanodonoghue, have you addressed this?
Yep

linaro	ssh://git.linaro.org/landing-teams/working/mbl/arm-trusted-firmware.git (fetch)
linaro	ssh://git.linaro.org/landing-teams/working/mbl/arm-trusted-firmware.git (push)

fb3bb64c856fc5ccb822a4781465bf33336022dc
grep -r FIP_ROM_OFFSET plat/imx/*
plat/imx/imx6/imx6ul_picopi/include/platform_def.h:#if defined(FIP_ROM_OFFSET)
plat/imx/imx6/imx6ul_picopi/include/platform_def.h:#define PICOPI_FIP_MMC_BASE		FIP_ROM_OFFSET

**Comment by bryanodonoghue** [comment]
Wrong link

https://git.linaro.org/landing-teams/working/mbl/arm-trusted-firmware.git/commit/?id=fb3bb64c856fc5ccb822a4781465bf33336022dc

https://git.linaro.org/landing-teams/working/mbl/arm-trusted-firmware.git/tree/plat/imx/imx6/imx6ul_picopi/include/platform_def.h?id=fb3bb64c856fc5ccb822a4781465bf33336022dc

**Comment by diego-sueiro** [comment]
Can you please run a full Jenkins build and LAVA tests?

**Comment by bryanodonoghue** [comment]
ok great

**Comment by bryanodonoghue** [comment]
jenkins.mbed-linux.arm.com/job/bod-warrior-dev-00/13/

**Comment by loverdeg-ep** [comment]
@ARMmbed/team-embeddedplanet

File: meta-freescale-3rdparty-mbl/conf/machine/imx7d-pico-mbl.conf [hunk 11]
```diff
@@ -45,7 +45,7 @@ KERNEL_IMAGETYPE = "fitImage"
 SPL_BINARY=""
 UBOOT_EXTLINUX = "0"
 UBOOT_CONFIG = ""
-UBOOT_MACHINE = "pico-pi-imx7d_defconfig"
+UBOOT_MACHINE = "pico-imx7d_bl33_defconfig"
```
Comment (diego-sueiro) [inline meta-freescale-3rdparty-mbl/conf/machine/imx7d-pico-mbl.conf]: @bryanodonoghue,

Can you please ensure that the following is printed during u-boot initialization?
`SEC0: RNG instantiated`

**Comment by bryanodonoghue** [comment]
Changed to
1. Only affect imx6ul 
2. Include license file information for u-boot GPL v2 + code
3. Verified imx7d is booting.

**Comment by bryanodonoghue** [comment]
http://jenkins.mbed-linux.arm.com/job/bod-warrior-dev-00/15

**Comment by bryanodonoghue** [comment]
https://jira.arm.com/browse/IOTMBL-2431 - captures subsequent imx6/imx7 integration task

**Pull request merged**